subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -51,56 +74,76 @@ public HierarchicalFileWatcherUpdater(FileWatcher watcher) {
     @Override
     public void changed(Collection<CompleteFileSystemLocationSnapshot> removedSnapshots, Collection<CompleteFileSystemLocationSnapshot> addedSnapshots) {
         removedSnapshots.forEach(snapshot -> {
-            ImmutableList<Path> previouslyWatchedRootsForSnapshot = watchedRootsForSnapshot.remove(snapshot.getAbsolutePath());
-            Multisets.removeOccurrences(shouldWatchDirectories, previouslyWatchedRootsForSnapshot);
+            trackedDirectoriesForSnapshot.removeAll(snapshot.getAbsolutePath());
         });
         addedSnapshots.forEach(snapshot -> {
             ImmutableList<Path> directoriesToWatch = WatchRootUtil.getDirectoriesToWatch(snapshot);
-            shouldWatchDirectories.addAll(directoriesToWatch);
-            watchedRootsForSnapshot.put(snapshot.getAbsolutePath(), directoriesToWatch);
+            trackedDirectoriesForSnapshot.putAll(snapshot.getAbsolutePath(), directoriesToWatch);
         });
         updateWatchedDirectories();
     }
 
     @Override
-    public void updateProjectRootDirectories(Collection<File> updatedProjectRootDirectories) {
-        Set<Path> rootPaths = updatedProjectRootDirectories.stream()
+    public void buildFinished() {
+        watchedRootProjectDirectoriesFromPreviousBuild.addAll(knownRootProjectDirectoriesFromCurrentBuild);
+        knownRootProjectDirectoriesFromCurrentBuild.clear();
+        updateWatchedDirectories();
+        watchedRootProjectDirectoriesFromPreviousBuild.removeIf(
+            rootProjectDirectory -> !watchedHierarchies.contains(rootProjectDirectory.path)
+        );
+    }
+
+    @Override
+    public void updateRootProjectDirectories(Collection<File> updatedRootProjectDirectories) {
+        Set<Path> rootPaths = updatedRootProjectDirectories.stream()
             .map(File::toPath)
             .map(Path::toAbsolutePath)
             .collect(Collectors.toSet());
-        projectRootDirectories.clear();
-        for (Path rootPathToWatch : WatchRootUtil.resolveRootsToWatch(rootPaths)) {
-            projectRootDirectories.put(rootPathToWatch, rootPathToWatch.toString() + File.separator);
-        }
-        LOGGER.info("Now considering {} as root directories to watch", projectRootDirectories.keySet());
+        Set<Path> newRootProjectDirectories = WatchRootUtil.resolveRootsToWatch(rootPaths);
+        LOGGER.info("Now considering {} as root directories to watch", newRootProjectDirectories);
+
+        knownRootProjectDirectoriesFromCurrentBuild.clear();
+        newRootProjectDirectories.stream()
+            .map(RootProjectDirectory::new)
+            .forEach(knownRootProjectDirectoriesFromCurrentBuild::add);
+        watchedRootProjectDirectoriesFromPreviousBuild.removeAll(knownRootProjectDirectoriesFromCurrentBuild);
+
         updateWatchedDirectories();
     }
 
     private void updateWatchedDirectories() {
```

Comment:
This is a bit weird for me somehow, having two methods with the same name, one with, the other without a parameter, calling each other. I think this one actually does two things, the first being collecting which hierarchies to watch, the second to actually update the watchers to watch these hierarchies. The other method with the parameter does only the actual updating. I think their names should reflect this relationship.